packages/@lwc/engine-core/src/framework/api.ts

```diff
@@ -791,8 +795,10 @@ function dc(
     // the new vnode key is a mix of idx and compiler key, this is required by the diffing algo
     // to identify different constructors as vnodes with different keys to avoid reusing the
     // element used for previous constructors.
-    data.key = `dc:${idx}:${data.key}`;
-    return c(sel, Ctor, data, children);
+    // Shallow clone is necessary here becuase VElementData may be shared across VNodes due to
+    // hoisting optimization.
+    const newData = { ...data, key: `dc:${idx}:${data.key}` };
```

Comment:
Ah, I was wondering what the key is for. If it's only to speed up DOM diffing, then what about removing it and relying on object equality? (Since objects are effectively immutable.) Or is the optimization still necessary?

Reply:
Keys are used to checking if vnodes are equal. it is primarily used whenever nodes are dynamic (if block, each block). In those cases, we can't use object equality as those VNodes can't be hoisted.

Reply:
keys are to denotate placeables for elements in the template, that's the way I think about them, and we use that information in multiple places to determine whether or not a vnode is the result of a particular placeable or another.

Reply:
another alternative here is to not use `...` due to the perf implications, and instead just use obj inheritance via `Object.create(data, { key: { value: `dc:${idx}:${data.key}` } })`. Considering that we never use `hasOwnProperty` with such object, we should be fine with this as well.

Reply:
again, this is not a critical path since dynamic components are almost never used, just saying! :)

Reply:
Hmm... I would personally prefer to keep `...` because it's idiomatic. Also, I'm guessing the perf implications are only from Locker, and [we disallow `lwc:dynamic` on the platform](https://lwc.dev/guide/reference#lwc%3Adynamic), so I'm assuming that this line of code will never run with Locker enabled?